gcloud-java-resourcemanager/src/main/java/com/google/gcloud/resourcemanager/testing/LocalResourceManagerHelper.java

```diff
@@ -0,0 +1,593 @@
+          project.setLabels(fullProject.getLabels());
+          break;
+        case "lifecycleState":
+          project.setLifecycleState(fullProject.getLifecycleState());
+          break;
+        case "name":
+          project.setName(fullProject.getName());
+          break;
+        case "parent":
+          project.setParent(fullProject.getParent());
+          break;
+        case "projectId":
+          project.setProjectId(fullProject.getProjectId());
+          break;
+        case "projectNumber":
+          project.setProjectNumber(fullProject.getProjectNumber());
+          break;
+      }
+    }
+    return project;
+  }
+
+  Response replace(Project project) {
+    Project oldProject = projects.get(project.getProjectId());
+    if (oldProject == null) {
+      // when possible, change this to 404 (#440)
+      return Error.PERMISSION_DENIED.response(
+          "Error when replacing " + project.getProjectId() + " because the project was not found.");
+    } else if (!oldProject.getLifecycleState().equals("ACTIVE")) {
+      return Error.FAILED_PRECONDITION.response("Error when replacing " + project.getProjectId()
+          + " because the lifecycle state was not ACTIVE.");
+    } else if (!Objects.equal(oldProject.getParent(), project.getParent())) {
+      return Error.INVALID_ARGUMENT.response(
+          "The server currently only supports setting the parent once "
+          + "and does not allow unsetting it.");
+    }
+    project.setLifecycleState("ACTIVE");
+    project.setProjectNumber(oldProject.getProjectNumber());
+    project.setCreateTime(oldProject.getCreateTime());
+    project.setParent(oldProject.getParent());
+    projects.put(project.getProjectId(), project);
+    try {
+      return new Response(HttpURLConnection.HTTP_OK, jsonFactory.toString(project));
+    } catch (IOException e) {
+      return Error.INTERNAL_ERROR.response(
+          "Error when serializing project " + project.getProjectId());
+    }
+  }
+
+  Response undelete(String projectId) {
+    Project project = projects.get(checkNotNull(projectId));
+    Response response;
+    if (project == null) {
+      // when possible, change this to 404 (#440)
+      response = Error.PERMISSION_DENIED.response(
+          "Error when undeleting " + projectId + " because the project was not found.");
+    } else if (!project.getLifecycleState().equals("DELETE_REQUESTED")) {
+      response = Error.FAILED_PRECONDITION.response("Error when undeleting " + projectId
+          + " because the lifecycle state was not DELETE_REQUESTED.");
+    } else {
+      project.setLifecycleState("ACTIVE");
+      response = new Response(HttpURLConnection.HTTP_OK, "{}");
+    }
+    return response;
+  }
+
+  private LocalResourceManagerHelper() {
+    InetSocketAddress addr = new InetSocketAddress(0);
+    try {
+      server = HttpServer.create(addr, 0);
+      port = server.getAddress().getPort();
+      server.createContext("/", new RequestHandler());
```

Comment:
Any reason not to specify `"/" + VERSION + "/projects/"` as the context?

Also, I would hope that in that case getting the projectId would be simple if
`getRequestURI().getPath()` would return a relative path to the context.